### PR TITLE
fix: ensure agent token is from latest build in middleware

### PR DESCRIPTION
**Pull request by f0ssel**
This applies the limit we currently use for some actions to all actions that use the WorkspaceAgent middleware.

### CODE REVIEW

**Review by deansheather — COMMENTED** [review]
LGTM, but needs a test that ensures other builds can't connect

**Comment by f0ssel** [comment]
@deansheather  Added the test

**Comment by spikecurtis** [comment]
I'm concerned that this is fragile because `GetWorkspaceAgentAndOwnerByAuthToken` will still happily return the agent for non-latest builds.  I realize that the middleware won't authorize non-latest builds and we use middleware for authentication pretty much everywhere right now, but I think we could make stronger guarantees and leave fewer sharp edges in our code if dropped `GetWorkspaceAgentAndOwnerByAuthToken` in favor of a single SQL query that takes the token and returns the agent and build (via joins) if it's the latest, and nothing if it's invalid or not the latest.  It would also reduce the number of DB roundtrips to boot.

**Comment by f0ssel** [comment]
Since I'm not the strongest SQL person, I'd like to propose the query before i go down the rabbit whole of changing all the interfaces and usages. Are you thinking something like this?
```sql
-- name: GetLatestWorkspaceAgentAndBuildAndOwnerByAuthToken :one
WITH LatestBuilds AS (
    SELECT
        workspaces.id AS workspace_id,
        MAX(workspace_builds.build_number) AS latest_build_number
    FROM workspaces
        INNER JOIN workspace_builds
        ON workspace_builds.workspace_id = workspaces.id
    WHERE workspaces.deleted = FALSE
    GROUP BY workspaces.id
)

SELECT
    sqlc.embed(workspace_agents),
    workspaces.id AS workspace_id,
    users.id AS owner_id,
    users.username AS owner_name,
    users.status AS owner_status,
    workspaces.template_id AS template_id,
    sqlc.embed(workspace_builds),
    array_cat(
        array_append(users.rbac_roles, 'member'),
        array_append(ARRAY[]::text[], 'organization-member:' || organization_members.organization_id::text)
    )::text[] as owner_roles,
    array_agg(COALESCE(group_members.group_id::text, ''))::text[] AS owner_groups
FROM users
    INNER JOIN workspaces
    ON workspaces.owner_id = users.id
    INNER JOIN workspace_builds
    ON workspace_builds.workspace_id = workspaces.id
    INNER JOIN LatestBuilds
    ON workspaces.id = LatestBuilds.workspace_id AND workspace_builds.build_number = LatestBuilds.latest_build_number
    INNER JOIN workspace_resources
    ON workspace_resources.job_id = workspace_builds.job_id
    INNER JOIN workspace_agents
    ON workspace_agents.resource_id = workspace_resources.id
    INNER JOIN organization_members
    ON organization_members.user_id = users.id
    LEFT JOIN group_members
    ON group_members.user_id = users.id
WHERE
    workspace_agents.auth_token = @auth_token
GROUP BY
    workspace_agents.id,
    workspaces.id,
    users.id,
    organization_members.organization_id,
    workspace_builds.build_number
ORDER BY
    workspace_builds.build_number DESC
LIMIT 1;
```
This generates the struct I think I would want, a row with both row.WorkspaceAgent and row.WorkspaceBuildTable. I believe this will return no rows when the auth token is not from the latest build. Does this seem like the direction you were thinking?

**Comment by spikecurtis** [comment]
Yeah, that's the kind of thing I was thinking.

BTW, I think the `LIMIT 1` is superfluous if the query is working correctly because we have a unique constraint on workspace_builds (workspace_id, build_number), which implies that exactly one build can be the "latest".

I also think it would read more clearly if you moved the `workspace_builds.build_number = LatestBuilds.latest_build_number` clause to `WHERE` rather than the join condition.  This is just style, but to me it's easier to understand that 

we're taking the workspace and looking up the latest build number, then only selecting rows where it's the latest build

vs

we're taking the workspace and then joining to a set of rows that is empty when it's not the latest

**Comment by f0ssel** [comment]
@spikecurtis I've spoken with @Emyrk on how to accomplish this query, and we've landed on the following changes:
1. `GetWorkspaceAgentAndOwnerByAuthToken` changed to `GetWorkspaceAgentAndLatestBuildByAuthToken` - rbac stuff was removed from query to simplify, array_agg was making things a mess. Since this query is only used here it's safe to change. 
2. Get rbac info from a call to `DB.GetAuthorizationUserRoles(dbauthz.AsSystemRestricted(ctx), row.Workspace.OwnerID)` in the middleware now. 

Tests are all passing now so I think this is good for review. I've added @Emyrk as a reviewer as well.

**Review by spikecurtis — APPROVED** [review]

File: coderd/database/gentest/models_test.go [hunk 1]
```diff
@@ -51,10 +51,10 @@ func TestViewSubsetTemplateVersion(t *testing.T) {
 	}
 }
 
-// TestViewSubsetWorkspaceBuild ensures WorkspaceBuildTable is a subset of WorkspaceBuild
+// TestViewSubsetWorkspaceBuild ensures WorkspaceBuild is a subset of WorkspaceBuild
 func TestViewSubsetWorkspaceBuild(t *testing.T) {
 	t.Parallel()
-	table := reflect.TypeOf(database.WorkspaceBuildTable{})
+	table := reflect.TypeOf(database.WorkspaceBuild{})
```
Comment (Emyrk) [inline coderd/database/gentest/models_test.go:57]: This change doesn't look too good. This is to ensure the view is always correct, so revert this change. The test should still pass.

**Review by Emyrk — CHANGES_REQUESTED** [review]
Let's get this test back, then it LGTM

File: coderd/database/gentest/models_test.go (hunk 1 above)
Comment (f0ssel) [reply coderd/database/gentest/models_test.go]: Good catch, went overboard on the find and replace

**Review by Emyrk — APPROVED** [review]
